pkg/supportbundle/supportbundle.go

```diff
@@ -0,0 +1,236 @@
+	"archive/zip"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"path/filepath"
+	"runtime"
+	"runtime/pprof"
+	"strings"
+	"time"
+
+	"github.com/grafana/agent/pkg/build"
+	"github.com/grafana/agent/pkg/config"
+	"github.com/grafana/agent/pkg/server"
+	"github.com/mackerelio/go-osstat/uptime"
+	"gopkg.in/yaml.v3"
+)
+
+// Bundle collects all the data that is exposed as a support bundle.
+type Bundle struct {
+	meta                  []byte
+	config                []byte
+	agentMetrics          []byte
+	agentMetricsInstances []byte
+	agentMetricsTargets   []byte
+	agentLogsInstances    []byte
+	heapBuf               *bytes.Buffer
+	goroutineBuf          *bytes.Buffer
+	blockBuf              *bytes.Buffer
+	mutexBuf              *bytes.Buffer
+	cpuBuf                *bytes.Buffer
+}
+
+// Metadata contains general runtime information about the current Agent.
+type Metadata struct {
+	BuildVersion string                 `yaml:"build_version"`
+	OS           string                 `yaml:"os"`
+	Architecture string                 `yaml:"architecture"`
+	Uptime       float64                `yaml:"uptime"`
+	Payload      map[string]interface{} `yaml:"payload"`
+}
+
+// Export gathers the information required for the support bundle.
+func Export(ctx context.Context, enabledFeatures []string, cfg config.Config, srvAddress string, dialContext server.DialContextFunc) (*Bundle, error) {
+	// The block profiler is disabled by default. Temporarily enable recording
+	// of all blocking events. Also, temporarily record all mutex contentions,
+	// and defer restoring of earlier mutex profiling fraction.
+	runtime.SetBlockProfileRate(1)
+	old := runtime.SetMutexProfileFraction(1)
+	defer func() {
+		runtime.SetBlockProfileRate(0)
+		runtime.SetMutexProfileFraction(old)
+	}()
+
+	// Gather runtime metadata.
+	ut, err := uptime.Get()
+	if err != nil {
+		return nil, err
+	}
+	m := Metadata{
+		BuildVersion: build.Version,
+		OS:           runtime.GOOS,
+		Architecture: runtime.GOARCH,
+		Uptime:       ut.Seconds(),
+		Payload:      map[string]interface{}{"enabled-features": enabledFeatures},
+	}
+	meta, err := yaml.Marshal(m)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal support bundle metadata: %s", err)
+	}
+	// Gather current configuration.
+	config, err := yaml.Marshal(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal config: %s", err)
+	}
+
+	var httpClient http.Client
+	httpClient.Transport = &http.Transport{DialContext: dialContext}
+	// Gather Agent's own metrics.
+	resp, err := httpClient.Get("http://" + srvAddress + "/metrics")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get internal Agent metrics: %s", err)
+	}
+	agentMetrics, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read internal Agent metrics: %s", err)
+	}
+
+	// Collect the Agent metrics instances and target statuses.
+	resp, err = httpClient.Get("http://" + srvAddress + "/agent/api/v1/metrics/instances")
```

Comment:
Do we need to check the tls flag and switch to HTTP is TLS is enabled?

Reply:
(I'm pretty sure this is using the in-memory HTTP client so TLS isn't required here)